<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="viewport" content="width=device-width, initial-scale=1">

<meta name="author" content="Stephen Bates" />


<title>Using the Digital Twin Test</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' || rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>



<link href="data:text/css;charset=utf-8,body%20%7B%0Abackground%2Dcolor%3A%20%23fff%3B%0Amargin%3A%201em%20auto%3B%0Amax%2Dwidth%3A%20700px%3B%0Aoverflow%3A%20visible%3B%0Apadding%2Dleft%3A%202em%3B%0Apadding%2Dright%3A%202em%3B%0Afont%2Dfamily%3A%20%22Open%20Sans%22%2C%20%22Helvetica%20Neue%22%2C%20Helvetica%2C%20Arial%2C%20sans%2Dserif%3B%0Afont%2Dsize%3A%2014px%3B%0Aline%2Dheight%3A%201%2E35%3B%0A%7D%0A%23header%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0A%23TOC%20%7B%0Aclear%3A%20both%3B%0Amargin%3A%200%200%2010px%2010px%3B%0Apadding%3A%204px%3B%0Awidth%3A%20400px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Aborder%2Dradius%3A%205px%3B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Afont%2Dsize%3A%2013px%3B%0Aline%2Dheight%3A%201%2E3%3B%0A%7D%0A%23TOC%20%2Etoctitle%20%7B%0Afont%2Dweight%3A%20bold%3B%0Afont%2Dsize%3A%2015px%3B%0Amargin%2Dleft%3A%205px%3B%0A%7D%0A%23TOC%20ul%20%7B%0Apadding%2Dleft%3A%2040px%3B%0Amargin%2Dleft%3A%20%2D1%2E5em%3B%0Amargin%2Dtop%3A%205px%3B%0Amargin%2Dbottom%3A%205px%3B%0A%7D%0A%23TOC%20ul%20ul%20%7B%0Amargin%2Dleft%3A%20%2D2em%3B%0A%7D%0A%23TOC%20li%20%7B%0Aline%2Dheight%3A%2016px%3B%0A%7D%0Atable%20%7B%0Amargin%3A%201em%20auto%3B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dcolor%3A%20%23DDDDDD%3B%0Aborder%2Dstyle%3A%20outset%3B%0Aborder%2Dcollapse%3A%20collapse%3B%0A%7D%0Atable%20th%20%7B%0Aborder%2Dwidth%3A%202px%3B%0Apadding%3A%205px%3B%0Aborder%2Dstyle%3A%20inset%3B%0A%7D%0Atable%20td%20%7B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dstyle%3A%20inset%3B%0Aline%2Dheight%3A%2018px%3B%0Apadding%3A%205px%205px%3B%0A%7D%0Atable%2C%20table%20th%2C%20table%20td%20%7B%0Aborder%2Dleft%2Dstyle%3A%20none%3B%0Aborder%2Dright%2Dstyle%3A%20none%3B%0A%7D%0Atable%20thead%2C%20table%20tr%2Eeven%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Ap%20%7B%0Amargin%3A%200%2E5em%200%3B%0A%7D%0Ablockquote%20%7B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Apadding%3A%200%2E25em%200%2E75em%3B%0A%7D%0Ahr%20%7B%0Aborder%2Dstyle%3A%20solid%3B%0Aborder%3A%20none%3B%0Aborder%2Dtop%3A%201px%20solid%20%23777%3B%0Amargin%3A%2028px%200%3B%0A%7D%0Adl%20%7B%0Amargin%2Dleft%3A%200%3B%0A%7D%0Adl%20dd%20%7B%0Amargin%2Dbottom%3A%2013px%3B%0Amargin%2Dleft%3A%2013px%3B%0A%7D%0Adl%20dt%20%7B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0Aul%20%7B%0Amargin%2Dtop%3A%200%3B%0A%7D%0Aul%20li%20%7B%0Alist%2Dstyle%3A%20circle%20outside%3B%0A%7D%0Aul%20ul%20%7B%0Amargin%2Dbottom%3A%200%3B%0A%7D%0Apre%2C%20code%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0Aborder%2Dradius%3A%203px%3B%0Acolor%3A%20%23333%3B%0Awhite%2Dspace%3A%20pre%2Dwrap%3B%20%0A%7D%0Apre%20%7B%0Aborder%2Dradius%3A%203px%3B%0Amargin%3A%205px%200px%2010px%200px%3B%0Apadding%3A%2010px%3B%0A%7D%0Apre%3Anot%28%5Bclass%5D%29%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Acode%20%7B%0Afont%2Dfamily%3A%20Consolas%2C%20Monaco%2C%20%27Courier%20New%27%2C%20monospace%3B%0Afont%2Dsize%3A%2085%25%3B%0A%7D%0Ap%20%3E%20code%2C%20li%20%3E%20code%20%7B%0Apadding%3A%202px%200px%3B%0A%7D%0Adiv%2Efigure%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0Aimg%20%7B%0Abackground%2Dcolor%3A%20%23FFFFFF%3B%0Apadding%3A%202px%3B%0Aborder%3A%201px%20solid%20%23DDDDDD%3B%0Aborder%2Dradius%3A%203px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Amargin%3A%200%205px%3B%0A%7D%0Ah1%20%7B%0Amargin%2Dtop%3A%200%3B%0Afont%2Dsize%3A%2035px%3B%0Aline%2Dheight%3A%2040px%3B%0A%7D%0Ah2%20%7B%0Aborder%2Dbottom%3A%204px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Apadding%2Dbottom%3A%202px%3B%0Afont%2Dsize%3A%20145%25%3B%0A%7D%0Ah3%20%7B%0Aborder%2Dbottom%3A%202px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Afont%2Dsize%3A%20120%25%3B%0A%7D%0Ah4%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23f7f7f7%3B%0Amargin%2Dleft%3A%208px%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Ah5%2C%20h6%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23ccc%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Aa%20%7B%0Acolor%3A%20%230033dd%3B%0Atext%2Ddecoration%3A%20none%3B%0A%7D%0Aa%3Ahover%20%7B%0Acolor%3A%20%236666ff%3B%20%7D%0Aa%3Avisited%20%7B%0Acolor%3A%20%23800080%3B%20%7D%0Aa%3Avisited%3Ahover%20%7B%0Acolor%3A%20%23BB00BB%3B%20%7D%0Aa%5Bhref%5E%3D%22http%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0Aa%5Bhref%5E%3D%22https%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0A%0Acode%20%3E%20span%2Ekw%20%7B%20color%3A%20%23555%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Edt%20%7B%20color%3A%20%23902000%3B%20%7D%20%0Acode%20%3E%20span%2Edv%20%7B%20color%3A%20%2340a070%3B%20%7D%20%0Acode%20%3E%20span%2Ebn%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Efl%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Ech%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Est%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Eco%20%7B%20color%3A%20%23888888%3B%20font%2Dstyle%3A%20italic%3B%20%7D%20%0Acode%20%3E%20span%2Eot%20%7B%20color%3A%20%23007020%3B%20%7D%20%0Acode%20%3E%20span%2Eal%20%7B%20color%3A%20%23ff0000%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Efu%20%7B%20color%3A%20%23900%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%20code%20%3E%20span%2Eer%20%7B%20color%3A%20%23a61717%3B%20background%2Dcolor%3A%20%23e3d2d2%3B%20%7D%20%0A" rel="stylesheet" type="text/css" />




</head>

<body>




<h1 class="title toc-ignore">Using the Digital Twin Test</h1>
<h4 class="author">Stephen Bates</h4>



<div id="summary" class="section level2">
<h2>Summary</h2>
<p>This document shows how to analyze genetic data with the Digital Twin Test (DTT) using the <code>digitaltwins</code> package.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(causalsnps)</code></pre></div>
</div>
<div id="create-a-synthetic-data-set" class="section level2">
<h2>Create a synthetic data set</h2>
<p>First, we create a synthetic data set which has two parts:</p>
<ol style="list-style-type: decimal">
<li>A an external data set of unrelated individuals.</li>
<li>A data set of parent-offspring trios.</li>
</ol>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#simulation parmaters</span>
n_ext &lt;-<span class="st"> </span><span class="dv">1000</span> <span class="co">#number of external observations</span>
n_trio &lt;-<span class="st"> </span><span class="dv">250</span> <span class="co">#number of trios</span>
p &lt;-<span class="st"> </span><span class="dv">500</span> <span class="co">#number of observed genetic variants</span>
chrome_width &lt;-<span class="st"> </span><span class="dv">50</span> <span class="co">#length of chromosome</span>
d &lt;-<span class="st"> </span><span class="kw">rep</span>(p <span class="op">/</span><span class="st"> </span>chrome_width, p) <span class="co">#genetic distance between sites</span>

k &lt;-<span class="st"> </span><span class="dv">5</span> <span class="co">#number of causal variants</span></code></pre></div>
<p>We will consider a single chromosome of length 50 Mb (roughly the length of chromosome 22), and take 500 genetic variants equally spaced across the chromosome. For simplicitly, we assume recombination occurs uniformly at random.</p>
<p>We begin with a matrix set of simulated haplotypes included in the <code>digitaltwins</code> package.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"> <span class="kw">data</span>(<span class="st">&quot;haps_matrix&quot;</span>)</code></pre></div>
<div id="external-data" class="section level4">
<h4>External data</h4>
<p>We take the first 2000 rows of <code>haps_matrix</code> as the haplotypes of the 1000 unrelated individuals (each individual has two corresponding haplotypes).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">external_haps &lt;-<span class="st"> </span>haps_matrix[<span class="dv">1</span><span class="op">:</span>(<span class="dv">2</span><span class="op">*</span>n_ext), ]
<span class="kw">dim</span>(external_haps)
<span class="co">#&gt; [1] 2000  500</span></code></pre></div>
</div>
<div id="trio-data" class="section level4">
<h4>Trio data</h4>
<p>Next, we use the remaining 1000 rows as the haplotypes of the parents for the 250 trios.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">parent_haps &lt;-<span class="st"> </span>haps_matrix[(<span class="dv">2</span><span class="op">*</span>n_ext <span class="op">+</span><span class="st"> </span><span class="dv">1</span>)<span class="op">:</span>(<span class="dv">2</span><span class="op">*</span>n_ext <span class="op">+</span><span class="st"> </span><span class="dv">4</span> <span class="op">*</span><span class="st"> </span>n_trio), ]
<span class="kw">dim</span>(parent_haps)
<span class="co">#&gt; [1] 1000  500</span></code></pre></div>
<p>We create the table of ancestries for the offspring haplotypes, and simulate the offspring haplotypes with the <code>generate_offspring</code> function.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">anc &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="dv">1</span><span class="op">:</span>(<span class="dv">4</span><span class="op">*</span>n_trio), <span class="dt">nrow =</span> <span class="dv">2</span><span class="op">*</span>n_trio, <span class="dt">byrow =</span> <span class="ot">TRUE</span>) <span class="co">#index of ancestors of haplotypes</span>
<span class="kw">print</span>(<span class="kw">dim</span>(anc))
<span class="co">#&gt; [1] 500   2</span>
<span class="kw">head</span>(anc)
<span class="co">#&gt;      [,1] [,2]</span>
<span class="co">#&gt; [1,]    1    2</span>
<span class="co">#&gt; [2,]    3    4</span>
<span class="co">#&gt; [3,]    5    6</span>
<span class="co">#&gt; [4,]    7    8</span>
<span class="co">#&gt; [5,]    9   10</span>
<span class="co">#&gt; [6,]   11   12</span></code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">set.seed</span>(<span class="dv">300</span>)
<span class="co"># use the &quot;generate_offspring&quot; function for each row of the ancestry table</span>
offspring_haps &lt;-<span class="st"> </span><span class="kw">mapply</span>(
  <span class="cf">function</span>(i, j) {<span class="kw">generate_offspring</span>(parent_haps[i, ], 
                                     parent_haps[i, ], 
                                     <span class="dt">d =</span> d)},
  anc[, <span class="dv">1</span>], anc[, <span class="dv">2</span>])
<span class="kw">dim</span>(offspring_haps)
<span class="co">#&gt; [1] 500 500</span></code></pre></div>
</div>
<div id="response-variable" class="section level4">
<h4>Response variable</h4>
<p>Lastly, we will simulate a response variable from a sparse linear regression model with 5 causal variants. The <code>haps_to_gen</code> function converts the matrix of haplotypes to a matrix of genotypes by adding rows 1 and 2, rows 3 and 4, etc.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#create the regression coefficients</span>
beta &lt;-<span class="st"> </span><span class="kw">rep</span>(<span class="dv">0</span>, p)
causal_variants &lt;-<span class="st"> </span><span class="kw">sample</span>(<span class="dv">1</span><span class="op">:</span>p, k)
beta[causal_variants] &lt;-<span class="st"> </span><span class="dv">1</span>

<span class="co">#sample the response variable from a sparse linear model</span>
Y_ext &lt;-<span class="st"> </span><span class="kw">haps_to_gen</span>(external_haps) <span class="op">%*%</span><span class="st"> </span>beta <span class="op">+</span><span class="st"> </span><span class="kw">rnorm</span>(n_ext)
Y_offspring &lt;-<span class="st"> </span><span class="kw">haps_to_gen</span>(offspring_haps) <span class="op">%*%</span><span class="st"> </span>beta <span class="op">+</span><span class="st"> </span><span class="kw">rnorm</span>(n_trio)</code></pre></div>
<p>We now have a population of unrelated haplotypes and parent-offspring trios, and so we will next turn to the Digital Twin Test.</p>
</div>
</div>
<div id="dtt-step-1-modeling-using-the-external-data" class="section level2">
<h2>DTT Step 1: Modeling using the External Data</h2>
<p>The first step of the Digital Twin Test is model fitting. The model fitting is done on the external data set of size 1000. This can be done using any fitting software that yields a linear predictor. We will use the <code>glmnet</code> package for sparse linear regression, tuning the model with cross-validation.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(glmnet)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">lasso_fit &lt;-<span class="st"> </span><span class="kw">cv.glmnet</span>(<span class="kw">haps_to_gen</span>(external_haps), Y_ext)
beta_hat &lt;-<span class="st"> </span><span class="kw">coef</span>(lasso_fit)
<span class="kw">length</span>(beta_hat) <span class="co">#Fitted linear predictor. First entry is an intercept.</span>
<span class="co">#&gt; [1] 501</span></code></pre></div>
</div>
<div id="dtt-step-2-inference-using-the-trio-data" class="section level2">
<h2>DTT Step 2: Inference using the Trio Data</h2>
<p>Next, we perform inference using the trio data. While our test uses a linear predictor, we emphasize that the validity of the test does not rely on the correctness of our model whatsoever. A better model fit will lead to higher power, but inference remains valid no matter the quality of the model fit.</p>
<p>We will take a group that contains a causal variant.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">test_region &lt;-<span class="st"> </span><span class="dv">1</span><span class="op">:</span><span class="dv">100</span>
<span class="kw">sum</span>(causal_variants <span class="op">%in%</span><span class="st"> </span>test_region) <span class="co">#number of causal variants in the test region</span>
<span class="co">#&gt; [1] 1</span></code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">p_value &lt;-<span class="st"> </span><span class="kw">linear_crt</span>(offspring_haps, parent_haps, anc, Y_offspring, 
                      <span class="kw">matrix</span>(beta_hat, <span class="dt">ncol =</span> <span class="dv">1</span>), 
                      <span class="dt">group =</span> test_region, <span class="dt">d =</span> d, <span class="dt">family =</span> <span class="st">&quot;gaussian&quot;</span>)
<span class="co">#&gt; Pre-computing forward-backward results... done.</span>
<span class="co">#&gt; Number of possible recombination events detected: 500.</span>
<span class="co">#&gt; Computing stats... </span>
<span class="co">#&gt; Estimating recombination events... </span>
<span class="co">#&gt; Number of recombination or mutation events in one repetition: 500.</span>
<span class="co">#&gt; Computing CRT p-values over 100 repetitions (using single core):</span>
<span class="co">#null p-value</span>
p_value 
<span class="co">#&gt; [1] 0.00990099</span></code></pre></div>
<p>We find that the p-value for this region is significant.</p>
<p>If we instead test a region that does not contain a causal variant, we will find a p-value that is not siginificant.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">test_region &lt;-<span class="st"> </span><span class="dv">400</span><span class="op">:</span><span class="dv">500</span>
<span class="kw">sum</span>(causal_variants <span class="op">%in%</span><span class="st"> </span>test_region) <span class="co">#number of causal variants in the test region</span>
<span class="co">#&gt; [1] 0</span></code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">p_value &lt;-<span class="st"> </span><span class="kw">linear_crt</span>(offspring_haps, parent_haps, anc, Y_offspring, 
                      <span class="kw">matrix</span>(beta_hat, <span class="dt">ncol =</span> <span class="dv">1</span>), 
                      <span class="dt">group =</span> test_region, <span class="dt">d =</span> d, <span class="dt">family =</span> <span class="st">&quot;gaussian&quot;</span>)
<span class="co">#&gt; Pre-computing forward-backward results... done.</span>
<span class="co">#&gt; Number of possible recombination events detected: 500.</span>
<span class="co">#&gt; Computing stats... </span>
<span class="co">#&gt; Estimating recombination events... </span>
<span class="co">#&gt; Number of recombination or mutation events in one repetition: 500.</span>
<span class="co">#&gt; Computing CRT p-values over 100 repetitions (using single core):</span>

<span class="co">#non-null p-value</span>
p_value 
<span class="co">#&gt; [1] 1</span></code></pre></div>
</div>



<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
