<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Using the Digital Twin Test â€¢ digitaltwins</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Using the Digital Twin Test">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">digitaltwins</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/first_steps.html">Getting Started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/first_steps.html">Using the Digital Twin Test</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/stephenbates19/digitaltwins">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Using the Digital Twin Test</h1>
                        <h4 class="author">Stephen Bates</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/stephenbates19/digitaltwins/blob/master/vignettes/first_steps.Rmd"><code>vignettes/first_steps.Rmd</code></a></small>
      <div class="hidden name"><code>first_steps.Rmd</code></div>

    </div>

    
    
<div id="summary" class="section level2">
<h2 class="hasAnchor">
<a href="#summary" class="anchor"></a>Summary</h2>
<p>This document shows how to analyze genetic data with the Digital Twin Test (DTT) using the <code>digitaltwins</code> package.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(digitaltwins)</code></pre></div>
</div>
<div id="create-a-synthetic-data-set" class="section level2">
<h2 class="hasAnchor">
<a href="#create-a-synthetic-data-set" class="anchor"></a>Create a synthetic data set</h2>
<p>First, we create a synthetic data set which has two parts:</p>
<ol style="list-style-type: decimal">
<li>A an external data set of unrelated individuals.</li>
<li>A data set of parent-offspring trios.</li>
</ol>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#simulation parmaters</span>
n_ext &lt;-<span class="st"> </span><span class="dv">1000</span> <span class="co">#number of external observations</span>
n_trio &lt;-<span class="st"> </span><span class="dv">250</span> <span class="co">#number of trios</span>
p &lt;-<span class="st"> </span><span class="dv">500</span> <span class="co">#number of observed genetic variants</span>
chrome_width &lt;-<span class="st"> </span><span class="dv">50</span> <span class="co">#length of chromosome</span>
d &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/rep.html">rep</a></span>(p <span class="op">/</span><span class="st"> </span>chrome_width, p) <span class="co">#genetic distance between sites</span>

k &lt;-<span class="st"> </span><span class="dv">5</span> <span class="co">#number of causal variants</span></code></pre></div>
<p>We will consider a single chromosome of length 50 Mb (roughly the length of chromosome 22), and take 500 genetic variants equally spaced across the chromosome. For simplicitly, we assume recombination occurs uniformly at random.</p>
<p>We begin with a matrix set of simulated haplotypes included in the <code>digitaltwins</code> package.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"> <span class="kw"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(<span class="st">"haps_matrix"</span>)</code></pre></div>
<div id="external-data" class="section level4">
<h4 class="hasAnchor">
<a href="#external-data" class="anchor"></a>External data</h4>
<p>We take the first 2000 rows of <code>haps_matrix</code> as the haplotypes of the 1000 unrelated individuals (each individual has two corresponding haplotypes).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">external_haps &lt;-<span class="st"> </span>haps_matrix[<span class="dv">1</span><span class="op">:</span>(<span class="dv">2</span><span class="op">*</span>n_ext), ]
<span class="kw"><a href="https://rdrr.io/r/base/dim.html">dim</a></span>(external_haps)
<span class="co">#&gt; [1] 2000  500</span></code></pre></div>
</div>
<div id="trio-data" class="section level4">
<h4 class="hasAnchor">
<a href="#trio-data" class="anchor"></a>Trio data</h4>
<p>Next, we use the remaining 1000 rows as the haplotypes of the parents for the 250 trios.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">parent_haps &lt;-<span class="st"> </span>haps_matrix[(<span class="dv">2</span><span class="op">*</span>n_ext <span class="op">+</span><span class="st"> </span><span class="dv">1</span>)<span class="op">:</span>(<span class="dv">2</span><span class="op">*</span>n_ext <span class="op">+</span><span class="st"> </span><span class="dv">4</span> <span class="op">*</span><span class="st"> </span>n_trio), ]
<span class="kw"><a href="https://rdrr.io/r/base/dim.html">dim</a></span>(parent_haps)
<span class="co">#&gt; [1] 1000  500</span></code></pre></div>
<p>We create the table of ancestries for the offspring haplotypes, and simulate the offspring haplotypes with the <code><a href="../reference/generate_offspring.html">digitaltwins::generate_offspring</a></code> function.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">anc &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span>(<span class="dv">1</span><span class="op">:</span>(<span class="dv">4</span><span class="op">*</span>n_trio), <span class="dt">nrow =</span> <span class="dv">2</span><span class="op">*</span>n_trio, <span class="dt">byrow =</span> <span class="ot">TRUE</span>) <span class="co">#index of ancestors of haplotypes</span>
<span class="kw"><a href="https://rdrr.io/r/base/print.html">print</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/dim.html">dim</a></span>(anc))
<span class="co">#&gt; [1] 500   2</span>
<span class="kw"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(anc)
<span class="co">#&gt;      [,1] [,2]</span>
<span class="co">#&gt; [1,]    1    2</span>
<span class="co">#&gt; [2,]    3    4</span>
<span class="co">#&gt; [3,]    5    6</span>
<span class="co">#&gt; [4,]    7    8</span>
<span class="co">#&gt; [5,]    9   10</span>
<span class="co">#&gt; [6,]   11   12</span></code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span>(<span class="dv">300</span>)
<span class="co"># use the "generate_offspring" function for each row of the ancestry table</span>
offspring_haps &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/mapply.html">mapply</a></span>(
  <span class="cf">function</span>(i, j) {digitaltwins<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/digitaltwins/man/generate_offspring.html">generate_offspring</a></span>(parent_haps[i, ], 
                                                 parent_haps[i, ], 
                                                 <span class="dt">d =</span> d)},
  anc[, <span class="dv">1</span>], anc[, <span class="dv">2</span>])
<span class="kw"><a href="https://rdrr.io/r/base/dim.html">dim</a></span>(offspring_haps)
<span class="co">#&gt; [1] 500 500</span></code></pre></div>
</div>
<div id="response-variable" class="section level4">
<h4 class="hasAnchor">
<a href="#response-variable" class="anchor"></a>Response variable</h4>
<p>Lastly, we will simulate a response variable from a sparse linear regression model with 5 causal variants. The <code><a href="../reference/haps_to_gen.html">digitaltwins::haps_to_gen</a></code> function converts the matrix of haplotypes to a matrix of genotypes by adding rows 1 and 2, rows 3 and 4, etc.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#create the regression coefficients</span>
beta &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/rep.html">rep</a></span>(<span class="dv">0</span>, p)
causal_variants &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/sample.html">sample</a></span>(<span class="dv">1</span><span class="op">:</span>p, k)
beta[causal_variants] &lt;-<span class="st"> </span><span class="dv">1</span>

<span class="co">#sample the response variable from a sparse linear model</span>
Y_ext &lt;-<span class="st"> </span>digitaltwins<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/digitaltwins/man/haps_to_gen.html">haps_to_gen</a></span>(external_haps) <span class="op">%*%</span><span class="st"> </span>beta <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span>(n_ext)
Y_offspring &lt;-<span class="st"> </span>digitaltwins<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/digitaltwins/man/haps_to_gen.html">haps_to_gen</a></span>(offspring_haps) <span class="op">%*%</span><span class="st"> </span>beta <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span>(n_trio)</code></pre></div>
<p>We now have a population of unrelated haplotypes and parent-offspring trios, and so we will next turn to the Digital Twin Test.</p>
</div>
</div>
<div id="dtt-step-1-modeling-using-the-external-data" class="section level2">
<h2 class="hasAnchor">
<a href="#dtt-step-1-modeling-using-the-external-data" class="anchor"></a>DTT Step 1: Modeling using the External Data</h2>
<p>The first step of the Digital Twin Test is model fitting. The model fitting is done on the external data set of size 1000. This can be done using any fitting software that yields a linear predictor. We will use the <code>glmnet</code> package for sparse linear regression, tuning the model with cross-validation.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(glmnet)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">lasso_fit &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/glmnet/man/cv.glmnet.html">cv.glmnet</a></span>(digitaltwins<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/digitaltwins/man/haps_to_gen.html">haps_to_gen</a></span>(external_haps), Y_ext)
beta_hat &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/stats/coef.html">coef</a></span>(lasso_fit)
<span class="kw"><a href="https://rdrr.io/r/base/length.html">length</a></span>(beta_hat) <span class="co">#Fitted linear predictor. First entry is an intercept.</span>
<span class="co">#&gt; [1] 501</span></code></pre></div>
</div>
<div id="dtt-step-2-inference-using-the-trio-data" class="section level2">
<h2 class="hasAnchor">
<a href="#dtt-step-2-inference-using-the-trio-data" class="anchor"></a>DTT Step 2: Inference using the Trio Data</h2>
<p>Next, we perform inference using the trio data. While our test uses a linear predictor, we emphasize that the validity of the test does not rely on the correctness of our model whatsoever. A better model fit will lead to higher power, but inference remains valid no matter the quality of the model fit.</p>
<p>We will take a group that contains a causal variant.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">test_region &lt;-<span class="st"> </span><span class="dv">1</span><span class="op">:</span><span class="dv">100</span>
<span class="kw"><a href="https://rdrr.io/r/base/sum.html">sum</a></span>(causal_variants <span class="op">%in%</span><span class="st"> </span>test_region) <span class="co">#number of causal variants in the test region</span>
<span class="co">#&gt; [1] 1</span></code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">p_value &lt;-<span class="st"> </span>digitaltwins<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/digitaltwins/man/linear_crt.html">linear_crt</a></span>(offspring_haps, parent_haps, anc, Y_offspring, 
                      <span class="kw"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span>(beta_hat, <span class="dt">ncol =</span> <span class="dv">1</span>), 
                      <span class="dt">group =</span> test_region, <span class="dt">d =</span> d, <span class="dt">family =</span> <span class="st">"gaussian"</span>)
<span class="co">#&gt; Pre-computing forward-backward results... done.</span>
<span class="co">#&gt; Number of possible recombination events detected: 500.</span>
<span class="co">#&gt; Computing stats... done.</span>
<span class="co">#&gt; Computing CRT p-values over 100 repetitions (using single core):</span>
<span class="co">#&gt; |----------------------------------------------------------------------------------------------------|</span>
<span class="co">#&gt; |</span>
<span class="co">#null p-value</span>
p_value 
<span class="co">#&gt; [1] 0.00990099</span></code></pre></div>
<p>We find that the p-value for this region is significant.</p>
<p>If we instead test a region that does not contain a causal variant, we will find a p-value that is not siginificant.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">test_region &lt;-<span class="st"> </span><span class="dv">400</span><span class="op">:</span><span class="dv">500</span>
<span class="kw"><a href="https://rdrr.io/r/base/sum.html">sum</a></span>(causal_variants <span class="op">%in%</span><span class="st"> </span>test_region) <span class="co">#number of causal variants in the test region</span>
<span class="co">#&gt; [1] 0</span></code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">p_value &lt;-<span class="st"> </span>digitaltwins<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/digitaltwins/man/linear_crt.html">linear_crt</a></span>(offspring_haps, parent_haps, anc, Y_offspring, 
                      <span class="kw"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span>(beta_hat, <span class="dt">ncol =</span> <span class="dv">1</span>), 
                      <span class="dt">group =</span> test_region, <span class="dt">d =</span> d, <span class="dt">family =</span> <span class="st">"gaussian"</span>)
<span class="co">#&gt; Pre-computing forward-backward results... done.</span>
<span class="co">#&gt; Number of possible recombination events detected: 500.</span>
<span class="co">#&gt; Computing stats... done.</span>
<span class="co">#&gt; Computing CRT p-values over 100 repetitions (using single core):</span>
<span class="co">#&gt; |----------------------------------------------------------------------------------------------------|</span>
<span class="co">#&gt; |</span>

<span class="co">#non-null p-value</span>
p_value 
<span class="co">#&gt; [1] 1</span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#summary">Summary</a></li>
      <li><a href="#create-a-synthetic-data-set">Create a synthetic data set</a></li>
      <li><a href="#dtt-step-1-modeling-using-the-external-data">DTT Step 1: Modeling using the External Data</a></li>
      <li><a href="#dtt-step-2-inference-using-the-trio-data">DTT Step 2: Inference using the Trio Data</a></li>
      </ul>
</div>
      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by <a href="https://web.stanford.edu/~stephen6/">Stephen Bates</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
