<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>DTT using a fixed linear statistic — linear_crt • digitaltwins</title>

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="DTT using a fixed linear statistic — linear_crt" />

<meta property="og:description" content="Returns p-values from the DTT using a linear feature importance measure." />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">digitaltwins</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/first_steps.html">Getting Started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/first_steps.html">Using the Digital Twin Test</a>
    </li>
  </ul>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/stephenbates19/digitaltwins">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>DTT using a fixed linear statistic</h1>
    <small class="dont-index">Source: <a href='https://github.com/stephenbates19/digitaltwins/blob/master/R/linear_crt.r'><code>R/linear_crt.r</code></a></small>
    <div class="hidden name"><code>linear_crt.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>Returns p-values from the DTT using a linear feature importance measure.</p>
    
    </div>

    <pre class="usage"><span class='fu'>linear_crt</span>(<span class='no'>H</span>, <span class='no'>H_parent</span>, <span class='no'>anc</span>, <span class='no'>Y</span>, <span class='no'>beta</span>, <span class='no'>group</span>, <span class='no'>d</span>, <span class='kw'>adjust</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>test_idx</span> <span class='kw'>=</span> <span class='fl'>1</span>:<span class='fu'><a href='https://rdrr.io/r/base/nrow.html'>nrow</a></span>(<span class='no'>H</span>), <span class='kw'>family</span> <span class='kw'>=</span> <span class='st'>"gaussian"</span>, <span class='kw'>n_reps</span> <span class='kw'>=</span> <span class='fl'>100</span>,
  <span class='kw'>genotypes</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>verbose</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>parallel</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>recomb_thresh</span> <span class='kw'>=</span> <span class='fl'>1e-08</span>, <span class='kw'>f_stat</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>H</th>
      <td><p>an (2n x p) matrix of the subject haplotypes.
It is assumed rows 1,2 belong to subject 1, rows 3,4 belong to subject 2, etc.</p></td>
    </tr>
    <tr>
      <th>H_parent</th>
      <td><p>and (n_2 x p) matrix of the parental haplotypes</p></td>
    </tr>
    <tr>
      <th>anc</th>
      <td><p>a (2n x 2) table of ancestries.
anc[i, 1] and and[i, 2] give the rows of H_parent corresponding to the
parents of row i of the haplotypes H</p></td>
    </tr>
    <tr>
      <th>Y</th>
      <td><p>observed responses, a vector of length n or a matrix of dimension (n x k)</p></td>
    </tr>
    <tr>
      <th>beta</th>
      <td><p>feature importance directions, fit on from an ind data set.
A vector of length p+1 or a matrix of dimension ((p+1) x k).
e.g. The output of get_beta_glment. The first coefficient is assumed to
be an intercept.</p></td>
    </tr>
    <tr>
      <th>group</th>
      <td><p>a list of of indices of the group to test. Should be a continuous region,
such as 10:20.</p></td>
    </tr>
    <tr>
      <th>d</th>
      <td><p>a vector of length p of genentic distances</p></td>
    </tr>
    <tr>
      <th>adjust</th>
      <td><p>a vector of length n or a matrix of dimension (n x k),
giving the contribution of the other chromosomes to the likelihood.
I.e. adjust = X <!-- %*% beta_hat where X is all other chromosomes and beta is -->
the fitted coefficients.</p></td>
    </tr>
    <tr>
      <th>test_idx</th>
      <td><p>a set of indices (between 1 and 2n) used to compute the test statistics</p></td>
    </tr>
    <tr>
      <th>family</th>
      <td><p>type of regression. Either "gaussian" or "binomial".
 If "guassian", correlation is used as a feature importance statistic. If "binomial",
logistic log-likelihood is used as a feature importance statistic.</p></td>
    </tr>
    <tr>
      <th>n_reps</th>
      <td><p>number of repitions of the CRT to carry out</p></td>
    </tr>
    <tr>
      <th>genotypes</th>
      <td><p>Defaults to TRUE. If FALSE, all rows are assumed to be independent
(i.e. no two haplotypes are from the same individual).</p></td>
    </tr>
    <tr>
      <th>verbose</th>
      <td><p>if TRUE, prints various diagnostic messages to console</p></td>
    </tr>
    <tr>
      <th>parallel</th>
      <td><p>requires doMC to be registered (default: FALSE)</p></td>
    </tr>
    <tr>
      <th>recomb_thresh</th>
      <td><p>threshold of probability of recombination events to ignore. Lower
values will have potentially higher power at increased computational cost.</p></td>
    </tr>
    <tr>
      <th>f_stat</th>
      <td><p>whether or not to use the f-statistic for continuous responses.
The f-statistic may be much slower for groups with many nonzero coefficients.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A vector of length k, a p-value for each column of Y.</p>
    

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>
          </ul>

  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by <a href='https://web.stanford.edu/~stephen6/'>Stephen Bates</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


